components/users/api/src/main/java/org/phenotips/groups/internal/UsersAndGroups.java

```diff
@@ -0,0 +1,304 @@
+     * @param searchUsers if true, includes users in result
+     * @param searchGroups if true, includes groups in result
+     * @return a json object containing all results found
+     */
+    public JSONObject search(String input, int maxResults, boolean searchUsers, boolean searchGroups)
+    {
+        String formattedInput = input.toLowerCase();
+        formattedInput = String.format(UsersAndGroups.INPUT_FORMAT, input);
+
+        JSONArray resultArray = new JSONArray();
+        try {
+            if (searchUsers) {
+                runUsersQuery(resultArray, UsersAndGroups.usersQueryString, formattedInput, maxResults);
+            }
+
+            if (searchGroups) {
+                runGroupsQuery(resultArray, UsersAndGroups.groupsQueryString, formattedInput, maxResults);
+            }
+        } catch (Exception ex) {
+            this.logger.error("Error in search ({})", input, ex.getMessage());
+        }
+
+        JSONObject result = new JSONObject();
+        result.put("matched", resultArray);
+        return result;
+    }
+
+    /**
+     * Searches for all users and groups. Result is returned as JSON
+     *
+     * @return a json object containing all results found
+     */
+    public JSONObject getAll()
+    {
+        JSONArray resultArray = new JSONArray();
+        try {
+            runUsersQuery(resultArray, UsersAndGroups.allUsersQueryString, "", 0);
+            runGroupsQuery(resultArray, UsersAndGroups.allGroupsQueryString, "", 0);
+        } catch (Exception ex) {
+            this.logger.error("Error in search for all users and groups ({})", ex.getMessage());
+        }
+
+        JSONObject result = new JSONObject();
+        result.put("results", resultArray);
+        return result;
+    }
+
+    private void runUsersQuery(JSONArray resultArray, String queryString, String formattedInput, int maxResults)
+        throws Exception
+    {
+        List<String> queryResult = runQuery(queryString, formattedInput, maxResults);
+        for (String userName : queryResult)
+        {
+            User user = this.userManager.getUser(userName);
+            List<AttachmentReference> attachmentRefs =
+                this.bridge.getAttachmentReferences((DocumentReference) user.getProfileDocument());
+            String avatarURL = "";
+            if (attachmentRefs.size() > 0) {
+                avatarURL = this.bridge.getAttachmentURL(attachmentRefs.get(0), true);
+            } else {
+                XWikiContext xcontext = this.xcontextProvider.get();
+                XWiki xwiki = xcontext.getWiki();
+                avatarURL = xwiki.getSkinFile("icons/xwiki/noavatar.png", xcontext);
+            }
+            String name = user.getName();
+            if (StringUtils.isBlank(name)) {
+                name = user.getUsername();
+            }
+            JSONObject o = createObject(user.getProfileDocument().toString(), name, avatarURL, USER);
+            resultArray.put(o);
```

Comment:
This is not very important, but if `resultArray` already has `maxResults` entries it will end up with `maxResults + 1` entries. Making this check before the loop iteration may actually save some work (if there are `maxResults` users any group search will be skipped when this method is called for the second time for groups)